packages/langium/src/lsp/workspace/workspace-manager.ts

```diff
@@ -0,0 +1,127 @@
+/******************************************************************************
+ * Copyright 2023 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { CancellationToken } from 'vscode-languageserver';
+import { interruptAndCheck } from '../../utils/promise-util.js';
+import { URI, UriUtils } from '../../utils/uri-util.js';
+import type { WorkspaceManager } from './../../workspace/workspace-manager.js';
+import type { WorkspaceFolder } from 'vscode-languageserver';
+import type { CoreServiceRegistryI } from '../../service-registry.js';
+import type { MutexLock } from '../../utils/promise-util.js';
+import type { BuildOptions, DocumentBuilder } from './../../workspace/document-builder.js';
+import type { LangiumDocument, LangiumDocuments } from './../../workspace/documents.js';
+import type { FileSystemNode, FileSystemProvider } from './../../workspace/file-system-provider.js';
+import type { LangiumSharedServices } from '../lsp-services.js';
+
+/**
+ * Langium Workspace manager that uses LSP workspace folders
+ */
+export class LSPWorkspaceManager implements WorkspaceManager {
```

Comment:
I need to intervene at this point, we need the `workspace initialization` also in CLI apps in case the language offers built-in libs or allows to include dependency packages.

Importing types, e.g. from `vscode-languageserver` does hopefully not hurt, but we need to inject our own `CancellationToken.None` in the non-lsp case.

Reply:
> Importing types, e.g. from `vscode-languageserver` does hopefully not hurt

It does hurt, because that the library we're hoping to exclude in the bundle if the adopter doesn't import from `langium/lsp`. However, I agree that the default workspace manager should still be available outside of the LSP export.

We should define our own `CancellationToken` interface in Langium.

Reply:
I can factor the cancellation token out in another set of changes, and see how that works out.

Reply:
>> Importing types, e.g. from vscode-languageserver does hopefully not hurt
>
> It does hurt, because that the library we're hoping to exclude in the bundle if the adopter doesn't import from langium/lsp

Just to make sure we're on the same page here: Does
`import type { WorkspaceFolder } from 'vscode-languageserver';`
have an impact on the bundling (tree shaking)?

Reply:
No, but `import { CancellationToken } from 'vscode-languageserver';` (which is in the file) has an impact.

Independent of whether the bundling can take care of that, I would much rather not import anything from `vscode-languageserver` outside of `langium/lsp`. That'll only invite trouble in the future.

Reply:
> No, but import { CancellationToken } from 'vscode-languageserver'; (which is in the file) has an impact.

That's for sure.

Reply:
Dito, although we could probably get away with just the types, it makes it fairly easy to import something that isn't a type by accident later.

Reply:
Last changes pull out and declare some of the necessary types from 'vscode-languageserver', which allows us to use CancelToken (and WorkspaceFolder) outside of the LSP context. The LSPWorkspaceManager is still present, but just extends upon the existing default, since that now provides ~90% of what we need anyways.

Some additional type changes were propagated using these internally defined types (CancellationToken & WorkspaceFolder) to some other services & example languages. We can still bend the type system as needed to determine where we want to switch between using ours vs. the vs-ls types, but I presume we'll assess in coming review passes.